Update GDAS INIT utility to optionally use fracoro data

ChunxiZhang-NOAA opened the pull request:
## DESCRIPTION OF CHANGES: 
Modified run_pre-v14.chgres.sh, run_v15.chgres.gfs.sh, run_v16.chgres.sh run_v14.chgres.sh, run_v15.chgres.sh and config files to make the new fracoro data as an option to be used by all run_*.chgres.sh.

## TESTS CONDUCTED: 
Testing is conducted for run_pre-v14.chgres.sh, run_v15.chgres.gfs.sh, run_v16.chgres.sh run_v14.chgres.sh, and run_v15.chgres.sh for both 'gfs' and 'gdas'. The selected testing dates are: 2022050606, 2020050606, 2019050606 and 2017050606. All tests passed. 

## DEPENDENCIES:
None

## DOCUMENTATION:
None

## ISSUE: 
Related to a global-workflow issue[#863](https://github.com/NOAA-EMC/global-workflow/issues/863).

ChunxiZhang-NOAA committed:
Add util/gdas_init/run_v17.chgres.sh to use fracoro data for all new UFS applications

GeorgeGayno-NOAA commented:
@ChunxiZhang-NOAA When I difference your new script with the v16 script, I see very few differences:

```
< # Run chgres using v17 netcdf history data as input.  These history
---
> # Run chgres using v16 netcdf history data as input.  These history
58,67d57
< if  [ ${CTAR} == 'C48' ] ; then
<    OCNRES='400'
< elif [ ${CTAR} == 'C96' ] ; then
<    OCNRES='100'
< elif [ ${CTAR} == 'C192' ] ; then
<    OCNRES='050'
< elif [ ${CTAR} == 'C384' ] || [ ${CTAR} == 'C768' ] || [ ${CTAR} == 'C1152' ]; then
<    OCNRES='025'
< fi
<
75,78c65,68
<  fix_dir_target_grid="${FIX_ORO}/${CTAR}.mx${OCNRES}_frac/fix_sfc"
<  mosaic_file_target_grid="${FIX_ORO}/${CTAR}.mx${OCNRES}_frac/${CTAR}_mosaic.nc"
<  orog_dir_target_grid="${FIX_ORO}/${CTAR}.mx${OCNRES}_frac"
<  orog_files_target_grid="oro_${CTAR}.mx${OCNRES}.tile1.nc","oro_${CTAR}.mx${OCNRES}.tile2.nc","oro_${CTAR}.mx${OCNRES}.tile3.nc","oro_${CTAR}.mx${OCNRES}.tile4.nc","oro_${CTAR}.mx${OCNRES}.tile5.nc","oro_${CTAR}.mx${OCNRES}.tile6.nc"
---
>  fix_dir_target_grid="${FIX_ORO}/${CTAR}/fix_sfc"
>  mosaic_file_target_grid="${FIX_ORO}/${CTAR}/${CTAR}_mosaic.nc"
>  orog_dir_target_grid="${FIX_ORO}/${CTAR}"
>  orog_files_target_grid="${CTAR}_oro_data.tile1.nc","${CTAR}_oro_data.tile2.nc","${CTAR}_oro_data.tile3.nc","${CTAR}_oro_data.tile4.nc","${CTAR}_oro_data.tile5.nc","${CTAR}_oro_data.tile6.nc"
```
Could you add logic to the existing v16 script to handle both cases instead of adding an entire new script? Suppose someone wants to use the fractional data with older data (v15 and previous)? Could logic be added to work with all the 'run' scripts?

GeorgeGayno-NOAA reviewed (changes requested):
See my latest comment.

ChunxiZhang-NOAA commented:
@GeorgeGayno-NOAA If use logic to control here, it also needs logic in the workflow. Currently, I use this v17 script in global workflow: https://github.com/NOAA-EMC/global-workflow/pull/1242/files. This new fracoro data should work for both fractional and non-fractional applications.

ChunxiZhang-NOAA committed:
Revert adding util/gdas_init/run_v17.chgres.sh and modify util/gdas_init/run_v16.chgres.sh to make all new UFS applications use fracoro data

ChunxiZhang-NOAA commented:
@GeorgeGayno-NOAA I added the changes as you suggested. It works with the global-workflow. Do I need to add the newly added switch "FRAC_ORO" in the config file to use UTIL in standalone mode?

GeorgeGayno-NOAA commented:
> @GeorgeGayno-NOAA I added the changes as you suggested. It works with the global-workflow. Do I need to add the newly added switch "FRAC_ORO" in the config file to use UTIL in standalone mode?

What you have is better, but it will only work if the user selects v16 data as input. Suppose someone wants to coldstart a case from 2020 **and** wants to use the fracoro data? In that case v15 data will be used as input to chgres and the `run_v15.chgres.sh` will be run. That v15 script (and the pre-v15 scripts) does not use the ORO_DIR and ORO_NAME variables. Is that your intent?

ChunxiZhang-NOAA commented:
> > @GeorgeGayno-NOAA I added the changes as you suggested. It works with the global-workflow. Do I need to add the newly added switch "FRAC_ORO" in the config file to use UTIL in standalone mode?
> 
> What you have is better, but it will only work if the user selects v16 data as input. Suppose someone wants to coldstart a case from 2020 **and** wants to use the fracoro data? In that case v15 data will be used as input to chgres and the `run_v15.chgres.sh` will be run. That v15 script (and the pre-v15 scripts) does not use the ORO_DIR and ORO_NAME variables. Is that your intent?

Yes, we only consider v16 and higher version. GFS v14 and v15 don't support fractional grid. And the chance to use new ORO data with GFSv14 and v15 is low.

GeorgeGayno-NOAA commented:
> > > @GeorgeGayno-NOAA I added the changes as you suggested. It works with the global-workflow. Do I need to add the newly added switch "FRAC_ORO" in the config file to use UTIL in standalone mode?
> > 
> > 
> > What you have is better, but it will only work if the user selects v16 data as input. Suppose someone wants to coldstart a case from 2020 **and** wants to use the fracoro data? In that case v15 data will be used as input to chgres and the `run_v15.chgres.sh` will be run. That v15 script (and the pre-v15 scripts) does not use the ORO_DIR and ORO_NAME variables. Is that your intent?
> 
> Yes, we only consider v16 and higher version. GFS v14 and v15 don't support fractional grid. And the chance to use new ORO data with GFSv14 and v15 is low.

Let me rephrase - I am not talking about initializing v14 or v15 runs. Rather, will you want to use archived v14 or v15 data to coldstart a v16 or higher run?

ChunxiZhang-NOAA commented:
> > > > @GeorgeGayno-NOAA I added the changes as you suggested. It works with the global-workflow. Do I need to add the newly added switch "FRAC_ORO" in the config file to use UTIL in standalone mode?
> > > 
> > > 
> > > What you have is better, but it will only work if the user selects v16 data as input. Suppose someone wants to coldstart a case from 2020 **and** wants to use the fracoro data? In that case v15 data will be used as input to chgres and the `run_v15.chgres.sh` will be run. That v15 script (and the pre-v15 scripts) does not use the ORO_DIR and ORO_NAME variables. Is that your intent?
> > 
> > 
> > Yes, we only consider v16 and higher version. GFS v14 and v15 don't support fractional grid. And the chance to use new ORO data with GFSv14 and v15 is low.
> 
> Let me rephrase - I am not talking about initializing v14 or v15 runs. Rather, will you want to use archived v14 or v15 data to coldstart a v16 or higher run?

Probably yes. Does it mean run_v14.chgres.sh and run_v15.chgres.sh also need to be modified?

GeorgeGayno-NOAA commented on util/gdas_init/run_v16.chgres.sh:
```diff
@@ -59,13 +59,30 @@ rm -fr $WORKDIR
 mkdir -p $WORKDIR
 cd $WORKDIR
 
```
Here is what I suggest. Move this logic that sets ORO_DIR and ORO_NAME to the 'config' file. Then at the bottom of the 'config' file, add an export of ORO_DIR and ORO_NAME.  Then, those variables would be available to all the chgres run scripts.

ChunxiZhang-NOAA replied:
@GeorgeGayno-NOAA Sounds a good plan. I will modify it as you suggested and test the code.

GeorgeGayno-NOAA replied:
@ChunxiZhang-NOAA I gave you some bad advice. Placing this new logic in the config file won't work for GDAS runs. Let me rework the logic and get back to you.

ChunxiZhang-NOAA replied:
@GeorgeGayno-NOAA OK, but I tested and it worked for GDAS runs, didn't test it for the latest UFS_UTILS version though.

ChunxiZhang-NOAA replied:
@GeorgeGayno-NOAA My GDAS test run logs on Hera: /scratch1/NCEPDEV/hwrf/save/Chunxi.Zhang/Fixfv3/latest/UFS_UTILS/util/gdas_init

GeorgeGayno-NOAA replied:
The problem for GDAS runs is that the hires and enkf members can be at different resolutions. For example, the default resolutions are C192/C96. When I tested your branch, both the hires and enkf members used C192. I think the logic for setting the orog data will need to be moved to its own script and then "sourced" from each 'run' script. I moved the following logic to its own script:

```
if [ ${MEMBER} == 'gdas' ] || [ ${MEMBER} == 'gfs' ] || [ ${MEMBER} == 'hires' ] ; then
  CTAR=${CRES_HIRES}
else
  CTAR=${CRES_ENKF}
fi

if [ "${FRAC_ORO:-"no"}" = "yes" ]; then
  if  [ ${CTAR} == 'C48' ] ; then
    OCNRES='400'
  elif [ ${CTAR} == 'C96' ] ; then
    OCNRES='100'
  elif [ ${CTAR} == 'C192' ] ; then
    OCNRES='050'
  elif [ ${CTAR} == 'C384' ] || [ ${CTAR} == 'C768' ] || [ ${CTAR} == 'C1152' ]; then
    OCNRES='025'
  fi
  ORO_DIR="${CTAR}.mx${OCNRES}_frac"
  ORO_NAME="oro_${CTAR}.mx${OCNRES}"
else
  ORO_DIR="${CTAR}"
  ORO_NAME="${CTAR}_oro_data"
fi
```

Then 'sourced' it from one of the 'run' scripts as follows:

```
 mkdir -p $OUTDIR
 mkdir -p $OUTDIR/INPUT

+source $UFS_DIR/util/gdas_init/set_fixed_files.sh
+
 cat << EOF > fort.41
```

That seemed to work. You may have a better way.

ChunxiZhang-NOAA replied:
That's odd. My C192/C192 tests on both Hera and wcoss2 were successful. Which platform did you use?

GeorgeGayno-NOAA replied:
> That's odd. My C192/C192 tests on both Hera and wcoss2 were successful. Which platform did you use?

Try a C192/C96 test.

ChunxiZhang-NOAA replied:
C192/C96 were successful on both Hera and wcoss2.

GeorgeGayno-NOAA replied:
I am looking at your Hera test - /scratch1/NCEPDEV/hwrf/save/Chunxi.Zhang/Fixfv3/latest/UFS_UTILS/util/gdas_init. You specified a hires of C192, and enkf of C96. But if you look in log.001, chgres is reading the C192 fix files:

```
- INPUT DATA FROM FV3 GAUSSIAN NETCDF FILE.
 - OPEN TARGET GRID MOSAIC FILE:
 /scratch1/NCEPDEV/hwrf/save/Chunxi.Zhang/Fixfv3/latest/UFS_UTILS/util/gdas_init
 /../../fix/orog/C192.mx050_frac/C192_mosaic.nc
 - TRACER NAME IN OUTPUT FILE sphum
```

ChunxiZhang-NOAA replied:
@GeorgeGayno-NOAA I see. Because variable 'member' has values like 1, 2, 3.... in run_*.chgres.sh, it is impossible to use CDUMP in the config file to get the correct 'CTAR'.  I made changes and put the definition of 'CTAR' back to the run_*.chgres.sh files. Then source the set_fixed_files.sh in the run_*.chgres.sh. Now tests for C192/C96 passed.

ChunxiZhang-NOAA committed:
Modify util/gdas_init/run_*.chgres.sh and config files to make fracoro data as an option

ChunxiZhang-NOAA committed:
Small changes to config file to reset to the default values

ChunxiZhang-NOAA commented:
@GeorgeGayno-NOAA Please review the code I modified as you suggested. I also moved the assignment of CTAR into config file. I tested for v16, v15,v14, v13, and all passed for both 'gfs' and 'gdas' runs.

GeorgeGayno-NOAA commented:
> @GeorgeGayno-NOAA Please review the code I modified as you suggested. I also moved the assignment of CTAR into config file. I tested for v16, v15,v14, v13, and all passed for both 'gfs' and 'gdas' runs.

I will review. And make sure your branch is up-to-date with develop.

ChunxiZhang-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/fixfv3_v17

GeorgeGayno-NOAA commented on util/gdas_init/config:
```diff
@@ -63,6 +63,8 @@ LEVS=65
 
 CDUMP=gdas
 
+FRAC_ORO=no
```
In the comments section, please define this variable.

GeorgeGayno-NOAA commented on util/gdas_init/run_v16retro.chgres.sh:
```diff
@@ -44,10 +44,8 @@ WORKDIR=${WORKDIR:-$OUTDIR/work.${MEMBER}}
 
 if [ ${MEMBER} == 'hires' ]; then
```
Be sure to update the namelist variables as you did in the other run scripts.

ChunxiZhang-NOAA committed:
Modify config file and run_*.chgres.sh

ChunxiZhang-NOAA committed:
Add util/gdas_init/set_fixed_files.sh to use the new ORO data

ChunxiZhang-NOAA committed:
Minor changes to config and run_v16.chgres.sh

GeorgeGayno-NOAA commented on util/gdas_init/set_fixed_files.sh:
```diff
@@ -0,0 +1,18 @@
+
+if [ "${FRAC_ORO:-"no"}" = "yes" ]; then
+  if  [ ${CTAR} == 'C48' ] ; then
+    OCNRES='400'
```
For C48, I believe OCNRES should be 500.

ChunxiZhang-NOAA replied:
Good catch! It should be 500. I have fixed it.

GeorgeGayno-NOAA commented:
@ChunxiZhang-NOAA Don't forget to update `run_v16retro.chgres.sh`.

GeorgeGayno-NOAA commented on util/gdas_init/set_fixed_files.sh:
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
```
Can you add a short prolog that explains what the script does?

ChunxiZhang-NOAA replied:
I've added it in the script as suggested.

ChunxiZhang-NOAA committed:
Minor changes to run_v16retro.chgres.sh and set_fixed_files.sh

ChunxiZhang-NOAA commented:
> @ChunxiZhang-NOAA Don't forget to update `run_v16retro.chgres.sh`.

@GeorgeGayno-NOAA I've updated and tested `run_v16retro.chgres.sh` as you suggested.

GeorgeGayno-NOAA commented:
@ChunxiZhang-NOAA I think this is ready for merge. Please update your branch with the latest updates from 'develop'.

GeorgeGayno-NOAA commented:
@KateFriedman-NOAA This allows the user to optionally choose the new 'frac' orog data - e.g., ./orog/C768.mx025_frac.

ChunxiZhang-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/fixfv3_v17

KateFriedman-NOAA commented:
FYI @aerorahul @WalterKolczynski-NOAA
